keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsStep.java

```diff
@@ -0,0 +1,154 @@
+package io.improbable.keanu.algorithms.mcmc;
+
+import io.improbable.keanu.algorithms.graphtraversal.VertexValuePropagation;
+import io.improbable.keanu.algorithms.mcmc.proposal.Proposal;
+import io.improbable.keanu.algorithms.mcmc.proposal.ProposalDistribution;
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.network.NetworkSnapshot;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+import lombok.Value;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+class MetropolisHastingsStep {
+
+    private static final double LOG_ZERO_PROBABILITY = Double.NEGATIVE_INFINITY;
+
+    private final ProposalDistribution proposalDistribution;
+    private final boolean useCacheOnRejection;
+    private final Map<Vertex, LambdaSection> affectedVerticesCache;
+    private final KeanuRandom random;
+
+    /**
+     * @param latentVertices       Vertices that are unknown/hidden variables
+     * @param proposalDistribution The proposal distribution
+     * @param useCacheOnRejection  True if caching values of the network such that recalculation isn't required
+     *                             on step rejection
+     * @param random               Source of randomness
+     */
+    MetropolisHastingsStep(List<? extends Vertex> latentVertices,
+                           ProposalDistribution proposalDistribution,
+                           boolean useCacheOnRejection,
+                           KeanuRandom random) {
+
+        this.proposalDistribution = proposalDistribution;
+        this.useCacheOnRejection = useCacheOnRejection;
+        this.random = random;
+        this.affectedVerticesCache = getVerticesAffectedByLatents(
```

Comment:
It could do with an explanation of why `useCacheOnRejection` is equivalent to `includeNonProbabilistic`

Reply:
Good idea!